cmd/forwarder/auth.go

```diff
@@ -0,0 +1,225 @@
+
+	if config.RedisUrl != "" {
+		// Parse redis db out of url
+		u, err := url.Parse(config.RedisUrl)
+		if err != nil {
+			return result, err
+		}
+
+		var db int
+		if u.Path != "" && u.Path != "/" {
+			db, err = strconv.Atoi(u.Path[1:len(u.Path)])
+			if err != nil {
+				return result, err
+			}
+		}
+
+		ui := u.User
+		password := ""
+		if ui != nil {
+			password, _ = ui.Password()
+		}
+		client := redis.NewClient(
+			&redis.Options{
+				Addr:     u.Host,
+				Password: password,
+				DB:       db,
+			},
+		)
+
+		// Refresh once at the start.
+		// Ignore errors here - if redis is down, we don't want the process to fail to start.
+		_, _ = refreshAuth(result, client, config.HmacKey, config.RedisKey, config.Tokens)
+
+		// Refresh forever
+		ticker := time.NewTicker(config.RefreshInterval)
+		go func() {
+			changed := false
+			for _ = range ticker.C {
+				_, err := refreshAuth(result, client, config.HmacKey, config.RedisKey, config.Tokens)
+				if err == nil {
+					pSuccesses.Inc(1)
+					if changed {
+						pChanges.Inc(1)
+					}
+				} else {
+					fmt.Printf("Unable to refresh credentials: %s", err)
+					pFailures.Inc(1)
+				}
+			}
+		}()
+	}
+
+	return result, err
+}
+
+// Refresh auth credentials.
+// Return true if credentials changed, false otherwise.
+func refreshAuth(ba *BasicAuth, client redis.Cmdable, hmacKey string, redisKey string, config string) (bool, error) {
+	// Start out using the strings from config
+	nba, err := NewBasicAuthFromString(config, hmacKey, ba.registry)
+	if err != nil {
+		return false, err
+	}
+
+	// Retrieve all secrets and construct a string in the format expected by BasicAuth
+	val := client.HGetAll(redisKey)
+	r, err := val.Result()
+	if err != nil {
+		return false, err
+	}
+
+	for k, v := range r {
+		var arr []Credential
+		err = json.Unmarshal([]byte(v), &arr)
+
+		if err != nil {
+			return false, err
+		}
+		nba.creds[k] = arr
+	}
+
+	// Swap the secrets if there are changes
+	if !reflect.DeepEqual(ba.creds, nba.creds) {
+		ba.Lock()
+		defer ba.Unlock()
+		ba.creds = nba.creds
+		return true, nil
+	}
+	return false, nil
+}
+
+// BasicAuth handles normal user/password Basic Auth requests, multiple
+// password for the same user and is safe for concurrent use.
+type BasicAuth struct {
+	sync.RWMutex
+	creds    map[string][]Credential
+	hmacKey  string
+	registry metrics.Registry
+}
+
+// NewBasicAuthFromString creates and populates a BasicAuth from the provided
+// credentials, encoded as a string, in the following format:
+// user:password|user:password|...
+func NewBasicAuthFromString(creds string, hmacKey string, registry metrics.Registry) (*BasicAuth, error) {
+	ba := NewBasicAuth(registry, hmacKey)
+
+	// If the string is empty, that's allowed.
+	if creds == "" {
+		return ba, nil
+	}
+
+	for _, u := range strings.Split(creds, "|") {
+		uparts := strings.SplitN(u, ":", 2)
+		if len(uparts) != 2 || len(uparts[0]) == 0 || len(uparts[1]) == 0 {
```

Comment:
Suggestion prefer: `uparts[0] == "" || uparts[1] == ""`

Reply:
I copied this wholesale from [here](https://github.com/heroku/authenticater/blob/master/basic_auth.go#L30-L32).

I tested this code out [here](https://play.golang.org/p/VSNxKKXAcyR) and I believe that using this suggestion will segfault if it encounters malformed input.